sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+##### Proposed flow
+
+For these reasons, I propose we extract the command-specific logic into its own class. That way, the the *Api* class will be exposed to the specific details of a command, how to package the request, how to make the call, what format the request should be in. This works to isolate the areas of volatility should we require to support different use cases for a command. 
+
+```python
+class ConjurClient()
+	def __init__(...):
+    self.api_client = ApiClient()
+    
+	def list(self, list_constraints: dict = None) -> dict:
+    list_api = ListApi(list_constraints)
+	  return self.list_api.execute(list_constraints)
+
+# interface common to all commands (list, variable, policy, etc)
+class CommandInterface()
+	def build_request()
+  def execute()
+  def format_request()
+  
+class ListApi(CommandInterface)
+	def __init__(api_client = None):
+    self.api_client = api_client
+
+	def build_request(list_constraints: dict = None)
+		# … validate …
+		# … build params for request
+		# … return built list request
+    
+	def format_request()
+		# … format request (JSON, string, etc)
+    
+  def execute(list_constraints: dict = None)
+		# builds the request
+    self.build_request(list_constraints)  
+    # makes the API call
+		response = self.api_client.call_api(…)
+    # formats response to List/JSON/etc
+		return self.format_request(response)
+  
+class ApiClient():
+  def __init__(...):
+     # …    
+
+  def call_api(...)
+  	# 1. build query and headers
+		# 2. escape data
+    # 3. perform request
+    # 4. return response
+
+```
+
+
+
+### Full proposed SDK flow
+
+![SDKReactorFlow](design/assets/SDKReactorFlow.png)
+
+At a high level,
+
+1. The user will initalize a Configuration and Credential object, *separate* from the Client
+
+   `config = conjur.Configuration(action=file)`, `creds = conjur.Credentials(action=keystore)`
+
+2. For each flow, Configuration and Credential classes will determine which flow to use and perform the relevant validations
+
+3. *Configuration* and *Credentials* objects will be returned to the user to be used when initalizing the Client
+
+4. The Client is initalized,
+
+   `client = conjur.ConjurClient(config, creds, debug=True, ...)`
+
+5. The user can now make a request,
+
+   `list = client.list()`
+
+6. Once a request is made, the *ConjurClient* makes a call to the command's associated API class, *ListAPI* which will handle all things list-specific; will build the request, make the request, and format the response
+
+7. The *ListAPI* will then make a request to the *APIClient* which will handle the escaping of parameters, build headers, decide to run in a secure or unsecure flow, etc.
+
+8. *APIClient* makes a request to the *Conjur server* and returns the response to be formatted.
+
+9. The formatted response is returned to the user.
+
+
+
+### Backwards compatibility
```

Comment:
I would add a sentence about the versioning. what will happen after these changes?